### PR TITLE
MayaTransformAPI

### PR DESCRIPTION
**Pull request by robthebloke**
Within AL_USDUtils, there is a new MayaTransformAPI class that will allow you to easily work with maya transforms in USD. Includes Python bindings.

### CODE REVIEW

**Comment by kxl-adsk** [comment]
We mostly reviewed the API as part of #158, so let's review CMake changes. Assigning @HamedSabri-adsk to review it.

File: test/lib/usd/utils/CMakeLists.txt [hunk 1]
```diff
@@ -33,3 +33,62 @@ mayaUsd_add_test(${TARGET_NAME}
...
+# sources
+# -----------------------------------------------------------------------------
+target_sources(${TARGET_NAME}
+    PRIVATE
+        main.cpp
+        test_MayaTransformAPI.cpp
+)
+add_definitions(
+  -DUSDUTILS_TEST_DATA="${CMAKE_CURRENT_SOURCE_DIR}/test_data"
+)
```
Comment (HamedSabri-adsk) [inline test/lib/usd/utils/CMakeLists.txt:51]: don't use `add_definitions`, instead use  `target_compile_definitions`

```
# -----------------------------------------------------------------------------
# compile configuration
# -----------------------------------------------------------------------------
target_compile_definitions(${TARGET_NAME}
    PRIVATE
        USDUTILS_TEST_DATA="${CMAKE_CURRENT_SOURCE_DIR}/test_data"
)
```
Please see the cmake section in the guidelines:
https://github.com/Autodesk/maya-usd/blob/dev/doc/codingGuidelines.md

File: test/lib/usd/utils/CMakeLists.txt [hunk 2]
```diff
@@ -33,3 +33,62 @@ mayaUsd_add_test(${TARGET_NAME}
...
+set(USDUTILS_PYTHON_TEST_NAME Python:MayaTransformAPI)
+add_test(
+  NAME ${USDUTILS_PYTHON_TEST_NAME}
+  COMMAND
+    python -m unittest discover -s ${CMAKE_CURRENT_SOURCE_DIR}
+)
+set_tests_properties(${USDUTILS_PYTHON_TEST_NAME}
+  PROPERTIES
+    ENVIRONMENT PYTHONPATH=${CMAKE_BINARY_DIR}/${INSTALL_DIR_SUFFIX}/usdutils:${CMAKE_BINARY_DIR}/${INSTALL_DIR_SUFFIX}/lib/usd/utils:$ENV{PYTHONPATH} 
+)
```
Comment (HamedSabri-adsk) [inline test/lib/usd/utils/CMakeLists.txt:94]: We shouldn't use `add_test` directly here. I see your comment `using older approach as mayaUsd_add_test wants to use mayapy`. What is the problem with mayapy?

Nit-pick: I would also remove the comment lines 81-83

File: test/lib/usd/utils/CMakeLists.txt [hunk 3]
```diff
@@ -33,3 +33,62 @@ mayaUsd_add_test(${TARGET_NAME}
...
+# -----------------------------------------------------------------------------
+# link libraries
+# -----------------------------------------------------------------------------
+target_link_libraries(${TARGET_NAME}
+    PRIVATE 
+        GTest::GTest
+        mayaUsdUtils
+        tf
+        sdf
+        ${Boost_PYTHON_LIBRARY}
```
Comment (HamedSabri-adsk) [inline test/lib/usd/utils/CMakeLists.txt:70]: use imported targets rather than variables: 

`Boost::python`

File: lib/usd/utils/CMakeLists.txt [hunk 4]
```diff
@@ -2,14 +2,20 @@ set(TARGET_NAME "mayaUsdUtils")
 
 add_library(${TARGET_NAME} SHARED)
 
+set(MAYAUSDUTILS_PYTHON_LIBRARY_NAME _${TARGET_NAME})
+set(MAYAUSDUTILS_PYTHON_LIBRARY_LOCATION ${CMAKE_INSTALL_PREFIX}/lib/python/${TARGET_NAME})
```
Comment (HamedSabri-adsk) [inline lib/usd/utils/CMakeLists.txt:6]: Nit-Pick: These two lines can go before line 3.

File: lib/usd/utils/CMakeLists.txt [hunk 5]
```diff
@@ -84,3 +92,75 @@ if(${CMAKE_SYSTEM_NAME} MATCHES "Windows")
...
+list(APPEND mayausdutils_python_source
+  module.cpp
+  wrapMayaTransformAPI.cpp
+)
+
+add_library(${MAYAUSDUTILS_PYTHON_LIBRARY_NAME}
+  SHARED
+    ${mayausdutils_python_source}
+)
+
```
Comment (HamedSabri-adsk) [inline lib/usd/utils/CMakeLists.txt:120]: don't use mayausdutils_python_source list. Use `target_sources`

```
# -----------------------------------------------------------------------------
# sources
# -----------------------------------------------------------------------------
target_sources(${MAYAUSDUTILS_PYTHON_LIBRARY_NAME}
    PRIVATE
        module.cpp
        wrapMayaTransformAPI.cpp
)
```

File: lib/usd/utils/CMakeLists.txt [hunk 6]
```diff
@@ -84,3 +92,75 @@ if(${CMAKE_SYSTEM_NAME} MATCHES "Windows")
...
+    ${UTILS_INCLUDE_LOCATION} 
+    ${MAYAUSDUTILS_INCLUDE_LOCATION} 
+    ${USD_INCLUDE_DIR}
+    )
+
+target_link_libraries(${MAYAUSDUTILS_PYTHON_LIBRARY_NAME}
+  ${TARGET_NAME}
+  ${python_LIBRARIES}
+  usdGeom
+)
```
Comment (HamedSabri-adsk) [inline lib/usd/utils/CMakeLists.txt:142]: It is important to specify the keyword ( PRIVATE, PUBLIC, INTERFACE ). It is unclear to me what the `build` and `usage` requirements are for this target?

See cmake code guidelines: 
https://github.com/Autodesk/maya-usd/blob/dev/doc/codingGuidelines.md

File: lib/usd/utils/CMakeLists.txt [hunk 7]
```diff
@@ -84,3 +92,75 @@ if(${CMAKE_SYSTEM_NAME} MATCHES "Windows")
...
+
+set_target_properties(${MAYAUSDUTILS_PYTHON_LIBRARY_NAME}
+  PROPERTIES
+    PREFIX "_"
+    SUFFIX ".so"
+    OUTPUT_NAME ${TARGET_NAME}
+    LIBRARY_OUTPUT_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/${TARGET_NAME}
+)
```
Comment (HamedSabri-adsk) [inline lib/usd/utils/CMakeLists.txt:150]: This code is not portable. Use `set_python_module_property`.

```
set_target_properties(${MAYAUSDUTILS_PYTHON_LIBRARY_NAME}
  PROPERTIES 
    OUTPUT_NAME ${TARGET_NAME}
    LIBRARY_OUTPUT_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}
)

set_python_module_property(${MAYAUSDUTILS_PYTHON_LIBRARY_NAME})
```

Does the PREFIX have to be `_`?  In other place when `set_python_module_property ` is called, prefix is set to empty.

https://github.com/Autodesk/maya-usd/blob/f34c6ed315ae509a06b3a054c2a38a1c65f870b9/cmake/utils.cmake#L364

File: lib/usd/utils/CMakeLists.txt [hunk 8]
```diff
@@ -84,3 +92,75 @@ if(${CMAKE_SYSTEM_NAME} MATCHES "Windows")
...
+endforeach()
+
+
+# install python library
+install(TARGETS ${MAYAUSDUTILS_PYTHON_LIBRARY_NAME}
+  LIBRARY
+  DESTINATION ${MAYAUSDUTILS_PYTHON_LIBRARY_LOCATION}
+  RUNTIME
+  DESTINATION ${MAYAUSDUTILS_PYTHON_LIBRARY_LOCATION}
+)
```
Comment (HamedSabri-adsk) [inline lib/usd/utils/CMakeLists.txt:166]: Follow the style please. Adding below comment blocks helps to read things better: 

```
# -----------------------------------------------------------------------------
# sources
# -----------------------------------------------------------------------------

# -----------------------------------------------------------------------------
# compiler configuration
# -----------------------------------------------------------------------------

# -----------------------------------------------------------------------------
# include directories
# -----------------------------------------------------------------------------

# -----------------------------------------------------------------------------
# link libraries
# -----------------------------------------------------------------------------

# -----------------------------------------------------------------------------
# install
# -----------------------------------------------------------------------------
```

File: lib/usd/utils/CMakeLists.txt [hunk 9]
```diff
@@ -84,3 +92,75 @@ if(${CMAKE_SYSTEM_NAME} MATCHES "Windows")
...
+
+set(PY_INIT_FILES
+  ${CMAKE_CURRENT_SOURCE_DIR}/__init__.py
+)
+
+# copy init files to build location for module tests
+foreach(INPUT_FILE ${PY_INIT_FILES})
+  string(REPLACE ${CMAKE_CURRENT_SOURCE_DIR} ${CMAKE_CURRENT_BINARY_DIR}/${TARGET_NAME} OUTPUT_FILE ${INPUT_FILE})
+  execute_process(COMMAND ${CMAKE_COMMAND} -E copy ${INPUT_FILE} ${OUTPUT_FILE})
+endforeach()
```
Comment (HamedSabri-adsk) [inline lib/usd/utils/CMakeLists.txt:108]: move all these after the call to `add_library`. 

Use `mayaUsd_copyFiles `. There are couple of issues with this kind of copy.

We want to copy resources before the target has been built, hence `PRE_BUILD` and `if and only something has changed`, hence `copy_if_different`.

`PRE_BUILD` COMMAND ${CMAKE_COMMAND} -E `copy_if_different`

https://github.com/Autodesk/maya-usd/blob/f34c6ed315ae509a06b3a054c2a38a1c65f870b9/cmake/utils.cmake#L222

File: test/lib/usd/utils/CMakeLists.txt (hunk 2 above)
Comment (murphyeoin) [reply test/lib/usd/utils/CMakeLists.txt:94]: Hi @HamedSabri-adsk - As this library doesn't require maya, we didn't want to have a dependency on maya in order to be able to run the tests. However this work was done quite a while ago, and it's now apparent that mayaUsd_add_test can be passed a COMMAND argument which doesn't require maya or mayapy, we can update

Comment (HamedSabri-adsk) [reply test/lib/usd/utils/CMakeLists.txt:94]: Thanks @murphyeoin for clarifying. Makes sense to me now.

File: test/lib/usd/utils/CMakeLists.txt [hunk 10]
```diff
@@ -93,8 +87,7 @@ mayaUsd_add_test(${TARGET_NAME}
 # -----------------------------------------------------------------------------
 
 set(USDUTILS_PYTHON_TEST_NAME Python:MayaTransformAPI)
-add_test(
-  NAME ${USDUTILS_PYTHON_TEST_NAME}
+mayaUsd_add_test(${USDUTILS_PYTHON_TEST_NAME}
   COMMAND
     python -m unittest discover -s ${CMAKE_CURRENT_SOURCE_DIR}
 )
```
Comment (HamedSabri-adsk) [inline test/lib/usd/utils/CMakeLists.txt:93]: @robthebloke  This code is not portable:

```
set_tests_properties(${USDUTILS_PYTHON_TEST_NAME}
  PROPERTIES
    ENVIRONMENT PYTHONPATH=${CMAKE_BINARY_DIR}/${INSTALL_DIR_SUFFIX}/usdutils:${CMAKE_BINARY_DIR}/${INSTALL_DIR_SUFFIX}/lib/usd/utils:$ENV{PYTHONPATH} 
)
```

Replace above with this:

```
set(python_path 
    ${CMAKE_BINARY_DIR}/${INSTALL_DIR_SUFFIX}/usdutils
    ${CMAKE_BINARY_DIR}/${INSTALL_DIR_SUFFIX}/lib/usd/utils
    $ENV{PYTHONPATH} 
)
separate_argument_list(python_path)

set_tests_properties(${USDUTILS_PYTHON_TEST_NAME}
  PROPERTIES
    ENVIRONMENT PYTHONPATH=${python_path}
)
```

Hmm, your python paths don't seem correct to me. Does this code even work?

File: test/lib/usd/utils/CMakeLists.txt [hunk 11]
```diff
@@ -71,7 +71,7 @@ target_link_libraries(${TARGET_NAME}
         mayaUsdUtils
         tf
         sdf
-        ${Boost_PYTHON_LIBRARY} #Boost::python
+        Boost::python
```
Comment (HamedSabri-adsk) [inline test/lib/usd/utils/CMakeLists.txt:74]: @robthebloke  Remove `Boost::python` or `${python_LIBRARIES}`. AFAIS, test_MayaTransformAPI test is written in C++, why are you linking against Python?

Remove `sdf, tf`. Why are you linking against these?

File: lib/usd/utils/CMakeLists.txt (hunk 4 above)
Comment (HamedSabri-adsk) [reply lib/usd/utils/CMakeLists.txt:4]: @robthebloke  Change 2 spaces to 4 spaces.

File: lib/usd/utils/CMakeLists.txt [hunk 12]
```diff
@@ -84,3 +90,87 @@ if(${CMAKE_SYSTEM_NAME} MATCHES "Windows")
             DESTINATION ${CMAKE_INSTALL_PREFIX}/lib OPTIONAL
     )
 endif()
+
+# -----------------------------------------------------------------------------
+# sources
+# -----------------------------------------------------------------------------
```
Comment (HamedSabri-adsk) [inline lib/usd/utils/CMakeLists.txt:96]: @robthebloke  Source block goes before `target_sources`
```
# -----------------------------------------------------------------------------
# sources
# -----------------------------------------------------------------------------
target_sources(${MAYAUSDUTILS_PYTHON_LIBRARY_NAME}
PRIVATE
  module.cpp
  wrapMayaTransformAPI.cpp
)
```

File: lib/usd/utils/CMakeLists.txt [hunk 13]
```diff
@@ -84,3 +90,87 @@ if(${CMAKE_SYSTEM_NAME} MATCHES "Windows")
...
+
+add_library(${MAYAUSDUTILS_PYTHON_LIBRARY_NAME} SHARED)
+
+target_sources(${MAYAUSDUTILS_PYTHON_LIBRARY_NAME}
+PRIVATE
+  module.cpp
+  wrapMayaTransformAPI.cpp
```
Comment (HamedSabri-adsk) [inline lib/usd/utils/CMakeLists.txt:103]: @robthebloke  wrapMayaTransformAPI has a dependency to **boost python** and you need to tell CMake to find it's module.

Add below after `add_library(${MAYAUSDUTILS_PYTHON_LIBRARY_NAME} SHARED)`

```
#------------------------------------------------------------------------------
# modules and definitions
#------------------------------------------------------------------------------
if(NOT DEFINED BOOST_ROOT)
    set(BOOST_ROOT ${PXR_USD_LOCATION})
endif()
if(MAYAUSD_DEFINE_BOOST_DEBUG_PYTHON_FLAG)
    set(Boost_USE_DEBUG_PYTHON ON)
endif()
find_package(Boost COMPONENTS python REQUIRED)
```

File: lib/usd/utils/CMakeLists.txt [hunk 14]
```diff
@@ -84,3 +90,87 @@ if(${CMAKE_SYSTEM_NAME} MATCHES "Windows")
...
+)
+
+# -----------------------------------------------------------------------------
+# include directories
+# -----------------------------------------------------------------------------
+target_include_directories(${MAYAUSDUTILS_PYTHON_LIBRARY_NAME} 
+  PUBLIC
+    ${UTILS_INCLUDE_LOCATION} 
+    ${MAYAUSDUTILS_INCLUDE_LOCATION} 
+    ${USD_INCLUDE_DIR}
```
Comment (HamedSabri-adsk) [inline lib/usd/utils/CMakeLists.txt:138]: @robthebloke  What are these?

```
${UTILS_INCLUDE_LOCATION} 
${MAYAUSDUTILS_INCLUDE_LOCATION} 
```

Get rid of them.

File: lib/usd/utils/CMakeLists.txt [hunk 15]
```diff
@@ -84,3 +90,87 @@ if(${CMAKE_SYSTEM_NAME} MATCHES "Windows")
...
+# -----------------------------------------------------------------------------
+# compiler configuration
+# -----------------------------------------------------------------------------
+target_compile_definitions(${MAYAUSDUTILS_PYTHON_LIBRARY_NAME}
+  PRIVATE
+    MFB_PACKAGE_NAME=${TARGET_NAME}
+    MFB_ALT_PACKAGE_NAME=${TARGET_NAME}
+    MFB_PACKAGE_MODULE=${TARGET_NAME}
+)
+
```
Comment (HamedSabri-adsk) [inline lib/usd/utils/CMakeLists.txt:130]: @robthebloke  does this branch even compiles anywhere?

You need to add this after :  `mayaUsd_compile_config(${MAYAUSDUTILS_PYTHON_LIBRARY_NAME})`

File: lib/usd/utils/CMakeLists.txt [hunk 16]
```diff
@@ -84,3 +90,87 @@ if(${CMAKE_SYSTEM_NAME} MATCHES "Windows")
...
+    )
+
+# -----------------------------------------------------------------------------
+# link libraries
+# -----------------------------------------------------------------------------
+target_link_libraries(${MAYAUSDUTILS_PYTHON_LIBRARY_NAME}
+  PRIVATE
+    ${TARGET_NAME}
+    ${python_LIBRARIES}
+    usdGeom
```
Comment (HamedSabri-adsk) [inline lib/usd/utils/CMakeLists.txt:148]: @robthebloke  Add `Boost::python` here now.   This the now proper target that really needs to link again python stuff.

Change `${python_LIBRARIES}`  to ${PYTHON_LIBRARIES}.  Make sure to do the same thing in bunch of other places where you added this.

File: test/lib/usd/utils/CMakeLists.txt [hunk 17]
```diff
@@ -93,8 +87,7 @@ mayaUsd_add_test(${TARGET_NAME}
 # -----------------------------------------------------------------------------
 
 set(USDUTILS_PYTHON_TEST_NAME Python:MayaTransformAPI)
-add_test(
-  NAME ${USDUTILS_PYTHON_TEST_NAME}
+mayaUsd_add_test(${USDUTILS_PYTHON_TEST_NAME}
   COMMAND
     python -m unittest discover -s ${CMAKE_CURRENT_SOURCE_DIR}
```
Comment (HamedSabri-adsk) [inline test/lib/usd/utils/CMakeLists.txt:92]: @robthebloke   Running python test from source code is not desirable. When these tests run they pollute the source code with bytecode ( *.pyc ) .

Merge dev into this branch and have a look at how UFE does it? We want to copy the any python codes and resources to the build directory and run the tests from there:

https://github.com/Autodesk/maya-usd/blob/819ce49d4bcfb50cbcd51f17adf2e2a3d5f30ea9/test/lib/ufe/CMakeLists.txt#L53

After the merge you will be able to set the working directory for the tests:

`WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}`

File: test/lib/usd/utils/CMakeLists.txt (hunk 11 above)
Comment (robthebloke) [reply test/lib/usd/utils/CMakeLists.txt:75]: Removing Tf, Boost::python, and PYTHON_LIBRARIES, results in unresolved externals.

**Pull request closed without merging**